icu4c/source/common/uloc.cpp

```diff
@@ -1465,6 +1465,25 @@ uloc_openKeywords(const char* localeID,
 static const char i_default[] = {'i', '-', 'd', 'e', 'f', 'a', 'u', 'l', 't'};
 #define I_DEFAULT_LENGTH UPRV_LENGTHOF(i_default)
 
+static int32_t _preprocessLocaleID(const char* inLocaleID, char* outBuffer, int32_t bufferCapacity, UErrorCode* err) {
+    // this function is to allow permissive parsing in locale IDs in general cases, where the locale ID can be in BCP
+    // 47 format, but with underbars as field delimiters
+    int32_t neededLength = static_cast<int32_t>(uprv_strlen(inLocaleID));
+    if (neededLength > bufferCapacity) {
```

Comment:
Why not copy into a CharString?

Reply:
You know, that actually _would_ be better.  I couldn't figure out a way to use CharString for the original problem I ended up creating PreflightingLocaleIDBuffer for, but I think it _will_ work here, and it's a lot less hassle.  Thanks for the suggestion.